### DISCUSSION TITLE
Plans for introducing SQL-like joins?

### DISCUSSION THREAD

**Discussion by dxnnymurphy** [perspective-dev/perspective#2858] [Q&A]
Hi, 
I have a found a post from a few years ago where a user asked about client side sql-like joins into tables:
https://github.com/finos/perspective/discussions/1509
The post states:
`Ad-hoc joins of already instantiated Tables are a a work-in-progress.`

Are there any updates on this? 

I am looking to  perform LEFT joins.
Thanks in advance!
Danny

**Comment by jordanskole** [comment]
Hey Danny, 

I do a lot of joining too. I don't know what platform you are on but it is possible to "left join" _kind of_ using a shared index. Here's how to do it: 

1. create an empty table using a schema, and an index (js/ts `baseTable = await worker.table(schema, { index })` 
2. load your first data set into the table `await baseTable.update(firstDataSet)`
3. load your second data set into the table `await baseTable.update(secondDataSet)`

This works _okay_ so long as you are fine collapsing both tables onto the index. It gets challenging however when you want to have a "tall" table and add extra information onto that. 


```typescript
const item_row_schema = {
  itemId: "integer", // or whatever 
  sharedId: "integer", // or whatever
  ...rest
}

const joining_row_schema = {
  sharedId: "integer", // or whatever 
  ...rest 
}
```

I am not sure if this part is possible.

**Reply by jordanskole** [reply]
Sorry, I should mention that this works in 2.10, I havent updated to 3+ yet

**Comment by dxnnymurphy** [comment]
Hey Jordan,
Unfortunately I am going for the sort of 'tall' table approach where there is no shared index. I will bare that in mind and maybe see if i can come up with some sort of shared index approach.
Thanks!

**Reply by jordanskole** [reply]
Yeah, I am currently stuck here too, would love to see true joins someday soon! I know that was in the 3.0+ release notes

**Reply by aszenz** [reply]
We are using duckdb-wasm for SQL like joins in the browser

The results are then piped to perspective

**Reply by jordanskole** [reply]
@aszenz I was actually considering that, how do you like it?

**Reply by aszenz** [reply]
It's good for pre data transformations before feeding the data into perspective , also helps with running totals and sums.
